src/realm/sync/noinst/migration_store.cpp

```diff
@@ -91,44 +131,70 @@ std::shared_ptr<realm::SyncConfig> MigrationStore::convert_sync_config(std::shar
     return flx_config;
 }
 
-void MigrationStore::migrate_to_flx(std::string rql_query_string)
+void MigrationStore::migrate_to_flx(std::string_view rql_query_string)
 {
+    REALM_ASSERT(m_state == MigrationState::NotMigrated);
     REALM_ASSERT(!rql_query_string.empty());
 
-    std::unique_lock lock{m_mutex};
-    m_query_string = rql_query_string;
-    m_state = MigrationState::Completed;
-    lock.unlock();
+    {
+        // Make sure the migration table has been initialized
+        if (!m_migration_table) {
+            REALM_ASSERT(load_data());
+        }
 
-    auto tr = m_db->start_write();
-    auto migration_store_obj = tr->get_table(m_migration_table)->get_object(0);
-    migration_store_obj.set(m_migration_query_str, m_query_string);
-    migration_store_obj.set(m_migration_state, int64_t(m_state));
-    migration_store_obj.set(m_migration_completed_at, Timestamp{std::chrono::system_clock::now()});
-    tr->commit();
+        std::unique_lock lock{m_mutex};
+        m_query_string = rql_query_string;
+        m_state = MigrationState::Migrated;
+
+        auto tr = m_db->start_write();
+        auto migration_table = tr->get_table(m_migration_table);
+        // This should be called in the non-migrated state, so the migration table should not exist
+        REALM_ASSERT(migration_table->is_empty());
+        auto migration_store_obj = migration_table->create_object();
+        migration_store_obj.set(m_migration_query_str, m_query_string);
+        migration_store_obj.set(m_migration_state, int64_t(m_state));
+        migration_store_obj.set(m_migration_completed_at, Timestamp{std::chrono::system_clock::now()});
+        tr->commit();
+    }
 
-    m_on_migration_state_changed(m_state);
+    m_on_migration_state_changed(MigrationState::Migrated);
 }
 
 void MigrationStore::cancel_migration()
 {
-    auto tr = m_db->start_write();
-    auto migration_table = tr->get_table(m_migration_table);
-    migration_table->clear();
-    tr->commit();
+    // Clear the migration state
+    std::unique_lock lock{m_mutex};
+    clear(std::move(lock));
 
-    {
-        std::lock_guard lock{m_mutex};
-        m_state = MigrationState::NotStarted;
+    m_on_migration_state_changed(MigrationState::NotMigrated);
+}
+
+void MigrationStore::clear(std::unique_lock<std::mutex> lock)
+{
+    // Make sure the migration table has been initialized
+    lock.unlock();
+    if (!m_migration_table) {
+        REALM_ASSERT(load_data());
```

Comment:
I think clear() can/should be called only after migration happened so the data should be loaded at this point. If anything, I would assert that m_migration_table exists.